File: Object-Oriented-Programming/Inheritance/inheritance.js
// JavaScript Implementation of Inheritance in OOP

class Person {
    // Base class representing a generic person
    constructor(name, age) {
        this.name = name;
        this.age = age;
    }
    
    introduce() {
        console.log(`Hello, my name is ${this.name} and I am ${this.age} years old.`);
    }
}

// Derived Class
class Teacher extends Person {
    constructor(name, age, subject) {
        super(name, age);
        this.subject = subject;
    }
    
    teach() {
        console.log(`I am teaching ${this.subject}.`);
    }
}

// Example Usage
const t1 = new Teacher("Alice", 35, "Mathematics");
t1.introduce();
t1.teach();

File: Object-Oriented-Programming/Inheritance/inheritance.cpp
// C# Implementation of Inheritance in OOP

using System;

// Base class representing a generic person
class Person {
    protected string Name;
    protected int Age;
    
    public Person(string name, int age) {
        Name = name;
        Age = age;
    }
    
    public void Introduce() {
        Console.WriteLine($"Hello, my name is {Name} and I am {Age} years old.");
    }
}

// Derived Class
class Teacher : Person {
    private string Subject;
    
    public Teacher(string name, int age, string subject) : base(name, age) {
        Subject = subject;
    }
    
    public void Teach() {
        Console.WriteLine($"I am teaching {Subject}.");
    }
}

// Example Usage
class Program {
    static void Main() {
        Teacher t1 = new Teacher("Alice", 35, "Mathematics");
        t1.Introduce();
        t1.Teach();
    }
}
